### PR TITLE
Feature/7571 two factor authentication

### PR DESCRIPTION
**Pull request by melsawy**

### CODE REVIEW

File: app/graph/mutations/generate_two_factor_backup_codes_mutation.rb [hunk 1]
```diff
@@ -0,0 +1,18 @@
+GenerateTwoFactorBackupCodesMutation = GraphQL::Relay::Mutation.define do
+  name 'GenerateTwoFactorBackupCodes'
+
+  input_field :id, !types.Int
+  
+  return_field :success, types.Boolean
+  return_field :codes, JsonStringType
+
+  resolve -> (_root, inputs, _ctx) {
+    user = User.where(id: inputs[:id]).last
```
Comment (caiosba) [inline app/graph/mutations/generate_two_factor_backup_codes_mutation.rb]: Please be sure that this user is the same as the current_api_user - we don't one user to request the 2FA code for another user :)

File: app/graph/mutations/user_two_factor_authentication_mutation.rb [hunk 2]
```diff
@@ -0,0 +1,21 @@
...
+  input_field :id, !types.Int
+  input_field :password, !types.String
+  input_field :qrcode, types.String
+  input_field :opt_required, types.Boolean
+
+  return_field :success, types.Boolean
+  return_field :user, UserType
+
+  resolve -> (_root, inputs, _ctx) {
+    user = User.where(id: inputs[:id]).last
```
Comment (caiosba) [inline app/graph/mutations/user_two_factor_authentication_mutation.rb]: Please be sure that this user is the same as the current_api_user - we don't one user to request the 2FA code for another user :)

File: app/models/concerns/user_two_factor_auth.rb [hunk 3]
```diff
@@ -0,0 +1,61 @@
+require 'active_support/concern'
```
Comment (caiosba) [inline app/models/concerns/user_two_factor_auth.rb]: Please fix the indentation in this file

File: app/models/concerns/user_two_factor_auth.rb [hunk 4]
```diff
@@ -0,0 +1,61 @@
...
+
+module UserTwoFactorAuth
+  extend ActiveSupport::Concern
+
+  included do
+  	attr_accessor :two_factor
+  	devise :two_factor_authenticatable, :two_factor_backupable,
+         :otp_secret_encryption_key => ENV['TWO_FACTOR_KEY']
```
Comment (caiosba) [inline app/models/concerns/user_two_factor_auth.rb]: As we discussed on Slack, please replace `ENV['TWO_FACTOR_KEY']` by `CONFIG['two_factor_key']` (and thus add `two_factor_key` to `config.yml`)

File: app/models/concerns/user_two_factor_auth.rb [hunk 5]
```diff
@@ -0,0 +1,61 @@
...
+	  def two_factor=(options)
+	  	errors = validate_two_factor(options)
+	  	raise errors.to_json unless errors.blank?
+  		begin
+    		self.otp_required_for_login = options[:otp_required]
+    		self.otp_secret = User.generate_otp_secret if options[:otp_required] == true
+    		self.skip_check_ability = true
+    		self.save!
+    	rescue StandardError => e
+    		raise e.message
```
Comment (caiosba) [inline app/models/concerns/user_two_factor_auth.rb]: Why are you rescuing if all you do in the block is to raise again?

Comment (melsawy) [reply app/models/concerns/user_two_factor_auth.rb]: Yes no need to raise, removed it

File: app/models/concerns/user_two_factor_auth.rb (hunk 4 above)
Comment (melsawy) [reply app/models/concerns/user_two_factor_auth.rb]: Fixed

File: app/models/concerns/user_two_factor_auth.rb (hunk 3 above)
Comment (melsawy) [reply app/models/concerns/user_two_factor_auth.rb]: Done

File: app/graph/mutations/user_two_factor_authentication_mutation.rb (hunk 2 above)
Comment (melsawy) [reply app/graph/mutations/user_two_factor_authentication_mutation.rb]: Fixed

File: app/graph/mutations/generate_two_factor_backup_codes_mutation.rb (hunk 1 above)
Comment (melsawy) [reply app/graph/mutations/generate_two_factor_backup_codes_mutation.rb]: Fixed

**Pull request merged**